activemodel/lib/active_model/serialization.rb

```diff
@@ -128,7 +128,9 @@ def serializable_hash(options = nil)
       return serializable_attributes(attribute_names) if options.blank?
 
       if only = options[:only]
-        attribute_names &= Array(only).map(&:to_s)
+        only = Array(only).map(&:to_s)
+        attribute_names &= only
+        attribute_names = attribute_names.in_order_of(:itself, only, filter: false)
```

Comment:
How about flipping the intersection

```suggestion
        attribute_names = Array(only).map(&:to_s) & attribute_names
```

Reply:
Smart!